x-pack/filebeat/input/netflow/integration_test.go

```diff
@@ -0,0 +1,388 @@
+						DataStream: &proto.DataStream{
+							Dataset: "netflow.log",
+						},
+						Source: integration.RequireNewStruct(t, map[string]interface{}{
+							"id":                    "netflow_integration_test",
+							"host":                  "localhost:6006",
+							"expiration_timeout":    "30m",
+							"queue_size":            2 * 4 * 1600,
+							"detect_sequence_reset": true,
+							"max_message_size":      "10KiB",
+						}),
+					},
+				},
+			},
+		},
+	}
+
+	healthyChan := make(chan struct{})
+	server := &mock.StubServerV2{
+		CheckinV2Impl: func(observed *proto.CheckinObserved) *proto.CheckinExpected {
+
+			if healthyChan != nil {
+				unitState, payload := extractStateAndPayload(observed, "input-unit-1")
+				if unitState == proto.State_HEALTHY {
+					if payload.streamStatusEquals("netflow-netflow.netflow-1e8b33de-d54a-45cd-90da-23ed71c482e2", map[string]interface{}{
+						"status": "HEALTHY",
+						"error":  "",
+					}) {
+						close(healthyChan)
+						healthyChan = nil
+					}
+				}
+			}
+
+			return &proto.CheckinExpected{
+				Units: allStreams,
+			}
+		},
+		ActionImpl: func(response *proto.ActionResponse) error {
+			return nil
+		},
+	}
+	if err := server.Start(); err != nil {
+		t.Fatalf("failed to start StubServerV2 server: %v", err)
+	}
+	defer server.Stop()
+
+	// It's necessary to change os.Args so filebeat.Filebeat() can read the
+	// appropriate args at beat.Execute().
+	initialOSArgs := os.Args
+	os.Args = []string{
+		"filebeat",
+		"-E", fmt.Sprintf(`management.insecure_grpc_url_for_testing="localhost:%d"`, server.Port),
+		"-E", "management.enabled=true",
+		"-E", "management.restart_on_output_change=true",
+		"-E", "logging.level=info",
+	}
+	defer func() {
+		os.Args = initialOSArgs
+	}()
+
+	beatCmd := filebeat.Filebeat()
+	beatRunErr := make(chan error)
+	go func() {
+		defer close(beatRunErr)
+		beatRunErr <- beatCmd.Execute()
+	}()
+
+	select {
+	case <-healthyChan:
+		break
+	case err := <-beatRunErr:
+		t.Fatalf("beat run err: %v", err)
+	case <-time.After(10 * time.Second):
+		t.Fatalf("timed out waiting for beat to become healthy")
+	}
+
+	registry := monitoring.GetNamespace("dataset").GetRegistry().GetRegistry("netflow_integration_test")
+
+	discardedEventsTotalVar, ok := registry.Get("discarded_events_total").(*monitoring.Uint)
+	require.True(t, ok)
+
+	receivedEventTotalVar, ok := registry.Get("received_events_total").(*monitoring.Uint)
+	require.True(t, ok)
+
+	udpAddr, err := net.ResolveUDPAddr("udp", "127.0.0.1:6006")
+	require.NoError(t, err)
+
+	conn, err := net.DialUDP("udp", nil, udpAddr)
+	require.NoError(t, err)
+
+	f, err := pcap.OpenOffline("testdata/pcap/ipfix_cisco.pcap")
+	require.NoError(t, err)
+	defer f.Close()
+
+	var totalBytes, totalPackets int
+	limiter := rate.NewLimiter(rate.Limit(10000), 1)
+
+	packetSource := gopacket.NewPacketSource(f, f.LinkType())
+	for pkt := range packetSource.Packets() {
+		for !limiter.Allow() {
```

Comment:
this would be a tight loop, with no breaker. I think using limiter.Wait would preferable.

Reply:
nope limiter.Wait is way heavier that Allow (it creates a timer and  then selects it) and my purpose is actually to make it a busy loop so I can send the next packet as fast as possible. Now maybe a better solution would be not to check if 1 packet is just allowed but batch them up, but I don't see the point of complicating this code as of now

Reply:
In only creates the timer if it needs to wait as far as I understand
https://cs.opensource.google/go/x/time/+/refs/tags/v0.5.0:rate/rate.go;l=282

Reply:
substituted `Allow` with `Wait`